src/ffdh.rs

```diff
@@ -0,0 +1,302 @@
+
+        attributes.push(attr_element!(
+            CKA_PRIME; OAFlags::AlwaysRequired | OAFlags::Unchangeable;
+            Attribute::from_bytes; val Vec::new()));
+        attributes.push(attr_element!(
+            CKA_BASE; OAFlags::AlwaysRequired | OAFlags::Unchangeable;
+            Attribute::from_bytes; val Vec::new()));
+        attributes.push(attr_element!(
+            CKA_VALUE; OAFlags::RequiredOnCreate
+            | OAFlags::SettableOnlyOnCreate | OAFlags::Unchangeable;
+            Attribute::from_bytes; val Vec::new()));
+
+        factory.data.finalize();
+
+        factory
+    }
+}
+
+impl ObjectFactory for FFDHPubFactory {
+    /// Creates an FFDH Public-Key Object from a template
+    ///
+    /// Validates that the provided attributes are consistent with the
+    /// factory via [ObjectFactory::default_object_create()]
+    ///
+    /// Additionally validates that the key is based on a well known
+    /// group based on safe primes
+    fn create(&self, template: &[CK_ATTRIBUTE]) -> Result<Object> {
+        let obj = self.default_object_create(template)?;
+        let _ = ffdh_groups::get_group_name(&obj)?;
+        Ok(obj)
+    }
+
+    fn get_data(&self) -> &ObjectFactoryData {
+        &self.data
+    }
+
+    fn get_data_mut(&mut self) -> &mut ObjectFactoryData {
+        &mut self.data
+    }
+}
+
+impl CommonKeyFactory for FFDHPubFactory {}
+
+impl PubKeyFactory for FFDHPubFactory {}
+
+/// The FFDH Private-Key Factory
+#[derive(Debug, Default)]
+pub struct FFDHPrivFactory {
+    data: ObjectFactoryData,
+}
+
+impl FFDHPrivFactory {
+    /// Initializes a new FFDH Private-Key factory
+    pub fn new() -> FFDHPrivFactory {
+        let mut factory: FFDHPrivFactory = Default::default();
+
+        factory.add_common_private_key_attrs();
+
+        let attributes = factory.data.get_attributes_mut();
+
+        attributes.push(attr_element!(
+            CKA_PRIME; OAFlags::RequiredOnCreate | OAFlags::Unchangeable;
+            Attribute::from_bytes; val Vec::new()));
+        attributes.push(attr_element!(
+            CKA_BASE; OAFlags::RequiredOnCreate | OAFlags::Unchangeable;
+            Attribute::from_bytes; val Vec::new()));
+        attributes.push(attr_element!(
+            CKA_VALUE; OAFlags::RequiredOnCreate
+            | OAFlags::SettableOnlyOnCreate | OAFlags::Unchangeable;
+            Attribute::from_bytes; val Vec::new()));
+        attributes.push(attr_element!(
+            CKA_VALUE_BITS;
+            OAFlags::SettableOnlyOnCreate | OAFlags::Unchangeable;
+            Attribute::from_ulong; val 0));
+
+        /* default to private */
+        let private = attr_element!(
+            CKA_PRIVATE; OAFlags::Defval | OAFlags::ChangeOnCopy;
+            Attribute::from_bool; val true);
+        match attributes.iter().position(|x| x.get_type() == CKA_PRIVATE) {
+            Some(idx) => attributes[idx] = private,
+            None => attributes.push(private),
+        }
```

Comment:
I think this code is verbosely in all the private key factories. Would it make sense to move it to separate helper function, as currently it needs a comment to be understandable on the first sight.